src/kOS.Safe/Execution/YiedFinishedThreadedDetector.cs

```diff
@@ -0,0 +1,89 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading;
+
+namespace kOS.Safe.Execution
+{
+    public abstract class YiedFinishedThreadedDetector : YieldFinishedDetector
+    {
+        private ManualResetEvent childThreadEvent;
+        private Thread childThread;
+        private Exception childException;
+
+        protected SafeSharedObjects shared;
+
+        public override void Begin(SafeSharedObjects sharedObj)
+        {
+            if (childThread != null)
+                throw new Exceptions.KOSException("Error calling Begin, childThread is not null.");
+
+            shared = sharedObj;
+
+            childThreadEvent = new ManualResetEvent(false);
+
+            ThreadInitialize(sharedObj);
+
+            childThread = new Thread(DoThread);
+            childThread.IsBackground = true;
+            childThread.Start();
+        }
+
+        public override bool IsFinished()
+        {
+            if (childThreadEvent.WaitOne(0))
+            {
+                childThread.Join();
+                if (childException == null)
+                {
+                    ThreadFinsh();
+                }
+                else
+                {
+                    // If it died due to a compile error, then we won't really be able to switch to program context
+                    // as was implied by calling Cpu.SwitchToProgramContext() up above.  The CPU needs to be
+                    // told that it's still in interpreter context, or else it fails to advance the interpreter's
+                    // instruction pointer and it will just try the "call run()" instruction again:
```

Comment:
This comment is confusing because it refers to supposedly calling Cpu.SwitchToProgramContext() "up above" but there's no such thing happening "above" this point in the file.  It's as if this comment might have been cut and pasted from somewhere else, and now here the "up above" doesn't make sense anymore.  At any rate, I don't know what it refers to.